FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1323 @@
+                                            .StackDataNode = ReducedPrecisionMode ? IREmit->_VFAbs(8, 8, StackMember.StackDataNode) :
+                                                                                    IREmit->_VAnd(16, 1, StackMember.StackDataNode, HelperNode),
+                                            .InterpretAsFloat = StackMember.InterpretAsFloat});
+        }
+
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+
+      case OP_F80CMPSTACK: {
+        const auto* Op = IROp->C<IROp_F80CmpStack>();
+        auto offset = Op->SrcStack;
+        const auto& [Valid1, StackMember1] = StackData.top();
+        const auto& [Valid2, StackMember2] = StackData.top(offset);
+
+        MigrateToSlowPathIf(Valid1 != StackSlot::VALID || Valid2 != StackSlot::VALID);
+
+        Ref CmpNode {};
+        if (SlowPath) {
+          Ref StackValue1 = LoadStackValueAtTop_Slow();
+          Ref StackValue2 = LoadStackValueAtOffset_Slow(offset);
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackValue1, StackValue2);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackValue1, StackValue2);
+          }
+        } else {
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackMember1.StackDataNode, StackMember2.StackDataNode);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackMember1.StackDataNode, StackMember2.StackDataNode);
+          }
+        }
+        IREmit->ReplaceUsesWithAfter(CodeNode, CmpNode, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+      case OP_F80STACKTEST: {
+        const auto* Op = IROp->C<IROp_F80StackTest>();
+        auto offset = Op->SrcStack;
+        const auto& [Valid, StackMember] = StackData.top(offset);
+
+        MigrateToSlowPathIf(Valid != StackSlot::VALID);
+
+        Ref CmpNode {};
+        Ref ZeroConst = IREmit->_VCastFromGPR(ReducedPrecisionMode ? 8 : 16, 8, GetConstant(0));
+
+        if (SlowPath) {
+          auto* StackValue = LoadStackValueAtOffset_Slow(offset);
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackValue, ZeroConst);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackValue, ZeroConst);
+          }
+        } else {
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackMember.StackDataNode, ZeroConst);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackMember.StackDataNode, ZeroConst);
+          }
+        }
+        IREmit->ReplaceUsesWithAfter(CodeNode, CmpNode, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+
+      case OP_F80CMPVALUE: {
+        const auto* Op = IROp->C<IROp_F80CmpValue>();
+        const auto& Value = CurrentIR.GetNode(Op->X80Src);
+        const auto& [Valid, StackMember] = StackData.top();
+
+        MigrateToSlowPathIf(Valid != StackSlot::VALID);
+
+        Ref CmpNode {};
+        if (SlowPath) {
```

Comment:
I think we want to pull out a helper `LoadStackValue()` that returns

```
SlowPath ? LoadStackValueAtTop_Slow() : StackMember.StackDataNode
```

Then use that throughout the file. It means we can totally deduplicate/collapse lots of if-statements, like this one here

Reply:
Lots of these indeed.